File: backend/src/services/upload.service.ts
import { Request } from "express";
import multer from "multer";
import path from "path";
import fs from "fs";
import { AppError } from "../middleware/error.middleware";

const MAX_FILE_SIZE = parseInt(process.env.MAX_FILE_SIZE || "5242880"); // 5MB default
const UPLOAD_DIR = process.env.UPLOAD_DIR || "./uploads/images";

// Ensure upload directory exists
if (!fs.existsSync(UPLOAD_DIR)) {
  fs.mkdirSync(UPLOAD_DIR, { recursive: true });
}

const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, UPLOAD_DIR);
  },
  filename: (req, file, cb) => {
    const uniqueSuffix = Date.now() + "-" + Math.round(Math.random() * 1e9);
    const ext = path.extname(file.originalname);
    cb(null, `image-${uniqueSuffix}${ext}`);
  },
});

const fileFilter = (req: Request, file: Express.Multer.File, cb: multer.FileFilterCallback) => {
  const allowedMimes = ["image/jpeg", "image/jpg", "image/png", "image/gif", "image/webp"];
  if (allowedMimes.includes(file.mimetype)) {
    cb(null, true);
  } else {
    cb(new AppError("Invalid file type. Only images are allowed.", 400));
  }
};

export const upload = multer({
  storage,
  limits: {
    fileSize: MAX_FILE_SIZE,
  },
  fileFilter,
});

export class UploadService {
  static getFileUrl(filename: string): string {
    return `/uploads/images/${filename}`;
  }

  static deleteFile(filename: string): void {
    const filePath = path.join(UPLOAD_DIR, filename);
    if (fs.existsSync(filePath)) {
      fs.unlinkSync(filePath);
    }
  }
}

File: backend-dotnet/Vietsov.Api/Services/UploadService.cs
using System.IO;

namespace Vietsov.Api.Services;

public interface IUploadService
{
    Task<string> UploadImageAsync(Stream fileStream, string fileName, string contentType);
    string GetFileUrl(string filename);
    Task DeleteFileAsync(string filename);
}

public class UploadService : IUploadService
{
    private readonly string _uploadPath;
    private readonly ILogger<UploadService> _logger;
    private readonly long _maxFileSize;
    private readonly string[] _allowedExtensions = { ".jpg", ".jpeg", ".png", ".gif", ".webp" };
    private readonly string[] _allowedMimeTypes = { "image/jpeg", "image/jpg", "image/png", "image/gif", "image/webp" };

    public UploadService(IConfiguration configuration, ILogger<UploadService> logger, IWebHostEnvironment environment)
    {
        _logger = logger;
        var uploadDir = configuration["FileUpload:Path"] ?? "uploads/images";
        _uploadPath = Path.Combine(environment.WebRootPath, uploadDir);
        _maxFileSize = long.Parse(configuration["FileUpload:MaxSize"] ?? "5242880"); // 5MB default

        // Ensure upload directory exists
        if (!Directory.Exists(_uploadPath))
        {
            Directory.CreateDirectory(_uploadPath);
        }
    }

    public async Task<string> UploadImageAsync(Stream fileStream, string fileName, string contentType)
    {
        // Validate file type
        var ext = Path.GetExtension(fileName).ToLowerInvariant();
        if (!_allowedExtensions.Contains(ext))
        {
            throw new InvalidOperationException("Invalid file type. Only images are allowed.");
        }

        if (!_allowedMimeTypes.Contains(contentType.ToLowerInvariant()))
        {
            throw new InvalidOperationException("Invalid file type. Only images are allowed.");
        }

        // Validate file size
        if (fileStream.Length > _maxFileSize)
        {
            throw new InvalidOperationException($"File size exceeds maximum allowed size of {_maxFileSize / 1024 / 1024}MB");
        }

        // Generate unique filename
        var uniqueSuffix = DateTime.UtcNow.Ticks + "-" + Guid.NewGuid().ToString("N")[..8];
        var newFileName = $"image-{uniqueSuffix}{ext}";
        var filePath = Path.Combine(_uploadPath, newFileName);

        // Save file
        using (var fileStreamOut = new FileStream(filePath, FileMode.Create))
        {
            await fileStream.CopyToAsync(fileStreamOut);
        }

        return newFileName;
    }

    public string GetFileUrl(string filename)
    {
        return $"/uploads/images/{filename}";
    }

    public async Task DeleteFileAsync(string filename)
    {
        var filePath = Path.Combine(_uploadPath, filename);
        if (File.Exists(filePath))
        {
            await Task.Run(() => File.Delete(filePath));
        }
    }
}

